lib/spack/spack/util/gcs.py

```diff
@@ -0,0 +1,216 @@
+            return self.bucket.get_blob(blob_path)
+        return None
+
+    def blob(self, blob_path):
+        if self.exists():
+            return self.bucket.blob(blob_path)
+        return None
+
+    def get_all_blobs(self, recursive=True, relative=True):
+        """Get a list of all blobs
+        Returns a list of all blobs within this bucket.
+
+        Args:
+            relative: If true (default), print blob paths
+                         relative to 'build_cache' directory.
+                      If false, print absolute blob paths (useful for
+                         destruction of bucket)
+        """
+        tty.debug('Getting GCS blobs... Recurse {0} -- Rel: {1}'.format(
+            recursive, relative))
+
+        if relative:
+            converter = self._relative_blob_name
+        else:
+            converter = str
+
+        if self.exists():
+            all_blobs = self.bucket.list_blobs(prefix=self.prefix)
+            blob_list = []
+
+            base_dirs = len(self.prefix.split('/')) + 1
+
+            for blob in all_blobs:
+                if not recursive:
+                    num_dirs = len(blob.name.split('/'))
+                    if num_dirs <= base_dirs:
+                        blob_list.append(converter(blob.name))
+                else:
+                    blob_list.append(converter(blob.name))
+
+            return blob_list
+
+    def _relative_blob_name(self, blob_name):
+        return os.path.relpath(blob_name, self.prefix)
+
+    def destroy(self, recursive=False, **kwargs):
+        """Bucket destruction method
+
+        Deletes all blobs within the bucket, and then deletes the bucket itself.
+
+        Uses GCS Batch operations to bundle several delete operations together.
+        """
+        from google.cloud.exceptions import NotFound
+        tty.debug("Bucket.destroy(recursive={0})".format(recursive))
+        try:
+            bucket_blobs = self.get_all_blobs(recursive=recursive, relative=False)
+            batch_size = 1000
+
+            num_blobs = len(bucket_blobs)
+            for i in range(0, num_blobs, batch_size):
+                with self.client.batch():
+                    for j in range(i, min(i + batch_size, num_blobs)):
+                        blob = self.blob(bucket_blobs[j])
+                        blob.delete()
+        except NotFound as ex:
+            tty.error("{0}, Could not delete a blob in bucket {1}.".format(
+                      ex, self.name))
+            sys.exit(1)
+
+
+class GCSBlob(object):
+    """GCS Blob object
+
+    Wraps some blob methods for spack functionality
+    """
+    def __init__(self, url, client=None):
+
+        self.url = url
+        if url.scheme != 'gs':
+            raise ValueError('Can not create GCS blob connection with scheme: {SCHEME}'
+                             .format(SCHEME=url.scheme))
+
+        if not client:
+            self.client = gcs_client()
+        else:
+            self.client = client
```

Comment:
This can be written more concisely as:
```suggestion
        self.client = client or gcs_client()
```

Reply:
👍🏻